psl-clojure/src/psl_clojure/core.clj

```diff
@@ -0,0 +1,401 @@
+                    (str "INFE: " (.getInfeasibility gr))
+                    (instance? WeightedGroundRule gr)
+                    (str "INCO: " (.getIncompatibility gr))))
+     (println (str "CLAS: " (.getSimpleName (.getClass gr))))
+     (cu/printlnw (str "STRI: " gr) 79)
+     (doseq [a (.getAtoms gr)]
+       (println (str "ATOM: " (.getValue a) ":" a)))))
+  ;; With rule name
+  ([ground-rules rule-name]
+   (ground-rules-print-summary
+    (ground-rules-by-name ground-rules rule-name))))
+
+(defn ground-rules-sample
+  "Return a sample of the ground rules up to a maximum size n."
+  [ground-rules n]
+  (cu/random-sample-n (seq ground-rules) n))
+
+(defn ground-rules-stratified-sample
+  "Return a sample of ground rules, with up to n of each rule."
+  [ground-rules n]
+  (for [rule-name (ground-rules-names ground-rules)
+        ground-rule (ground-rules-sample
+                     (ground-rules-by-name ground-rules rule-name)
+                     n)]
+    ground-rule))
+
+(defn rules-info
+  "Return info about rules."
+  [model]
+  (for [k (.getRules model)]
+    (if (instance? UnweightedLogicalRule k)
+      {:kind "constraint" :name (.getName k)}
+      {:kind "compatibility" :name (.getName k) :weight (.getWeight k)})))
+
+(defn model-new
+  "Return a new PSLModel associated with the given DataStore."
+  [data-store]
+  (Model. "" data-store))
+
+(defn model-print
+  "Print the given model."
+  [model]
+  (println (.toString model)))
+
+(defn mpe-inference
+  "Call mpeInference on the inference app."
+  [inference-app]
+  (log/info "inference:: ::starting")
+  (let [result (.mpeInference inference-app)]
+    (log/info "inference:: ::done")
+    result))
+
+(defn open-db
+  "Return an open PSL database."
+  ([datastore model parts-to-read part-to-write preds-to-close]
+     {:pre [(not-any? nil? [datastore model parts-to-read part-to-write preds-to-close])]}
+     (let [parts-to-read (into-array Partition parts-to-read)
+           preds-to-close (HashSet.
+                           (for [pname preds-to-close]
+                             (.getPredicate model pname)))]
+       (.getDatabase datastore part-to-write preds-to-close parts-to-read)))
+
+  ;; No predicates to close
+  ([datastore model parts-to-read part-to-write]
+     {:pre [(not-any? nil? [datastore model parts-to-read part-to-write])]}
+     (let [parts-to-read (into-array Partition parts-to-read)]
+       (.getDatabase datastore part-to-write parts-to-read)))
+
+  ;; No predicates to close or write partition
+  ([datastore model parts-to-read]
+     {:pre [(not-any? nil? [datastore model parts-to-read])]}
+     (let [parts-to-read (into-array Partition parts-to-read)]
+       (.getDatabase datastore (first parts-to-read) parts-to-read))))
+
+(defn p
+  "Get a predicate object from the PSL model"
+  [model pred-name]
+  (.getPredicate model pred-name))
+
+;;; =============== Functions for handling PSL partitions ====================
+
+(defn partition-copy-atoms
```

Comment:
Previous conversation:
```
eriq-augustine 7 days ago  Owner
General moving around of atoms is pretty strange.
We do have the ability to move atoms just to the write partition, but that is really just for special cases like lazy grounding:
https://github.com/eriq-augustine/psl/blob/develop/psl-core/src/main/java/org/linqs/psl/database/Database.java#L215

Is there a special use case for Clojure (or your project)?

alexmemory 4 days ago
Yes, I needed it for my project and it seemed possibly useful generally. Could it get a user in trouble?
```

Yeah, it can be dangerous.
The most important place that comes to mind is the database cache (and the atom managers).
All atoms that are fetched by the database get cached so that only one copy of an atom actually exists in memory (so we never have inconsistent states).
But moving around atoms can cause its theoretical type to be different (eg if it was moved to/from a write partition).